aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,127 @@
+             "to slow (for loop and stack) implementation");
+  TORCH_CHECK(std::all_of(op.schema().returns().begin(),
+                          op.schema().returns().end(),
+                          [] (const Argument& arg) { return arg.type() == TensorType::get(); }),
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We could not generate a fallback.");
+  TORCH_CHECK(num_returns == 1,
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We do not yet support operations with multiple returns.");
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatched(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
+  auto batch_sizes = ArrayRef<int64_t>(some_sizes.begin(), some_sizes.begin() + num_batch_dims);
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Populate `num_batches` number of torch::jit::Stack, one for each computation.
+  std::vector<torch::jit::Stack> unbatched_stacks(num_batches);
+  auto pushToEachStack = [&](const auto& ivalue) {
+    for (auto& stack : unbatched_stacks) {
+      torch::jit::push(stack, ivalue);
+    }
+  };
+  auto batched_tensor_inputs_pos_iter = batched_tensor_inputs_position.begin();
+	auto input_physical_views_iter = input_physical_views.begin();
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
+    if (idx != *batched_tensor_inputs_pos_iter) {
+      // ivalue isn't a BatchedTensor
+      pushToEachStack(ivalue);
+      continue;
+    }
+    // ivalue is a BatchedTensor
+    const auto& physical_view_for_ivalue = *input_physical_views_iter;
+    for (int64_t linear_idx = 0; linear_idx < num_batches; ++linear_idx) {
+      auto index = computeIndex(linear_idx, batch_sizes);
```

Comment:
Instead of computing the index every time, I could modify the index in-place and add 1 to it. Is this worth it? In terms of performance it looks like it is more efficient

Reply:
I don't think I care too much, either way.

Reply:
I'll stick with the current approach (re-compute the index vector) for now and optimize later if necessary